rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,406 @@
+---
+authors: Nic Klaassen (nic@goteleport.com)
+state: draft
+---
+
+# RFD 136 - Modern Signature Algorithms
+
+## Required Approvers
+
+TODO
+
+## What
+
+Teleport should support modern key types and signature algorithms, currently
+only RSA2048 keys are supported with the PKCS#1 v1.5 signature scheme.
+This applies to CA keys and client (user/host) keys, but each can/will be
+addressed individually.
+
+## Why
+
+Modern algorithms like ECDSA and Ed25519 offer better security properties with
+smaller keys that are faster to generate and sign with.
+Some of the more restrictive security policies are starting to reject RSA2048
+(e.g. [RHEL 8's FUTURE policy](https://access.redhat.com/articles/3642912)).
+
+## Details
+
+### Algorithms
+
+These algorithms are being considered for support:
+
+#### RSA
+
+Private key sizes: 2048, 3072, 4096
+
+Signature algorithms: PKCS#1 v1.5, (maybe PSS for TLS and JWT? SSH does not support it)
+
+Digest/hash algorithms: SHA512 for SSH, SHA256 for TLS
+
+Considerations:
+
+* RSA2048 is the current default and deviating from it by default may break
+  compatibility with third-party components and protocols
+* RSA has the most widespread support among all protocols
+* Certain database protocols only support RSA client certs
+  * <https://docs.snowflake.com/en/user-guide/key-pair-auth#step-2-generate-a-public-key>
+* If we must continue to support RSA, we might as well support larger key sizes,
+  3072 and 4096-bit are the most commonly used and supported by e.g. GCP KMS.
+* golang.org/x/crypto/ssh always uses SHA512 hash with RSA public keys
```

Comment:
This is not true. It defaults to SHA-2/512, but there are some cases where it sill uses SHA-1 (see `LegacySigner` in Teleport and Go/crypto).

Reply:
True, we can (and do) override this, SHA512 is just the default for RSA keys. But as far as I can tell, currently all user and host SSH certificates are signed with a SHA512 hash, and we fall back to SHA1 only with user keys for signing ssh auth requests, does that sound mostly accurate @jakule ?

Reply:
We fall back to SHA-1 for OpenSSH nodes that fail to authenticate with SHA-2. I was just pointing out that "always" is a strong word in this case.

Reply:
Yes, in general, we keep SHA-2 signed keys in the DB and return them to users (`tsh`, WebUI, etc). All the magic that happens with SHA-1 signing happens internally in Teleport. I just wanted to point out that we do have a few places that use SHA-1.